patches/net/minecraft/client/multiplayer/ClientLevel.java.patch

```diff
@@ -17,6 +17,21 @@
  
      public void handleBlockChangedAck(int p_233652_) {
          this.blockStatePredictionHandler.endPredictionsUpTo(p_233652_, this);
+@@ -157,10 +_,14 @@
+     @Override
+     public boolean setBlock(BlockPos p_233643_, BlockState p_233644_, int p_233645_, int p_233646_) {
+         if (this.blockStatePredictionHandler.isPredicting()) {
++            // Neo: Record and store a snapshot in the prediction so that BE data can be restored if the break is denied.
++            var snapshot = net.neoforged.neoforge.common.util.BlockSnapshot.create(this.dimension(), this, p_233643_);
```

Comment:
It may be a larger patch, but wouldn't it make more sense to implement this in `retainKnownServerState`?

Reply:
No, that would be impossible, since `setBlock` has already been called by the time control flow reaches `retainKnownServerState`.